File: JavaScript/MazeObjects/Cell.js
/**
 * Initializes a new instance of the cell class
 * @param  {int} x_coordinate [The position of the cell relative to the x-axis]
 * @param  {int} y_coordinate [The position of the cell relative to the y-axis]
 * @param  {int} width        [The thickness of the cell]
 * @constructor
 */
function Cell (x_coordinate, y_coordinate, width) {
  this.x_coordinate = x_coordinate;
  this.y_coordinate = y_coordinate;
  this.width = width;
  this.visited = false;
  this.Walls = {};
  this.BuildWalls();
}

/**
 * Initializes each wall of the cell
 */
Cell.prototype.BuildWalls = function () {
  let x = this.x_coordinate * this.width;
  let y = this.y_coordinate * this.width;
  let top = new CellWall([x, y],
    [x + this.width, y]);
  let bottom = new CellWall([x, y + this.width],
    [x + this.width, y + this.width])
  let right = new CellWall([x + this.width, y],
    [x + this.width, y + this.width]);
  let left = new CellWall([x, y],
    [x, y + this.width]);

  this.Walls = {"top": top, "bottom": bottom, "right": right, "left": left};
};

/**
 * Draws the cell onto the canvas
 */
Cell.prototype.Draw = function () {
  for (var wall in this.Walls) {
    if (this.Walls[wall].show) {
      stroke(255);
      this.Walls[wall].Draw();
    }
    else{
      noStroke();
      noFill();
    }
  }

  if (this.visited){
    noStroke();
    fill(102, 51, 153, 128);
    rect(this.x_coordinate * this.width,
      this.y_coordinate * this.width , this.width, this.width);
    }
  };

/**
 * Gets the index a cell in the grid based on its x and y coordinates
 * @param  {int} x_coordinate   The position of the cell relative to the x-axis
 * @param  {int} y_coordinate   The position of the cell relative to the y-axis
 * @param  {int} grid_row_count The number of rows present in the grid
 * @return {int}                The index of the cell in the grid array
 */
Cell.prototype.GetNeighborIndex = function (x_coordinate, y_coordinate, grid_row_count) {
  if ((x_coordinate < 0) || (y_coordinate < 0) || (x_coordinate > grid_row_count - 1)
        || (y_coordinate > grid_row_count - 1)){
          return - 1;
        }

    return x_coordinate + y_coordinate * grid_row_count;
};

/**
 * Gets the next neighbor of the current cell
 * @param  {Array} grid            The collection of cells
 * @param  {Number} [grid_width=10] The number of columns in the grid
 * @return {Cell}                 The next cell to be visited
 */
Cell.prototype.GetNeighbor = function (grid, grid_width = 10) {
  let neighbor_indexes = [
    this.GetNeighborIndex(this.x_coordinate, this.y_coordinate + 1, grid_width),
    this.GetNeighborIndex(this.x_coordinate, this.y_coordinate - 1, grid_width),
    this.GetNeighborIndex(this.x_coordinate + 1, this.y_coordinate, grid_width),
    this.GetNeighborIndex(this.x_coordinate - 1, this.y_coordinate, grid_width)
  ];

  let neighbors = [];

  for (index of neighbor_indexes) {
    if ((index != -1) && (!grid[index].visited)){
      neighbors.push(grid[index]);
    }
  }

  return random(neighbors);
};

/**
 * Draws a rectangle that is the same dimensions as the cell
 * @param  {Number} red   Red value
 * @param  {Number} green Green value
 * @param  {Number} blue  Blue value
 */
Cell.prototype.Highlight = function (red, green, blue, alpha = 128) {
  fill(red, green, blue, alpha);
  rect(this.x_coordinate * this.width, this.y_coordinate * this.width,
    this.width, this.width);
};

/**
 * Removes the wall between the current cell and its neighbor
 * @param  {Cell} neighbor The neighboring cell object
 */
Cell.prototype.RemoveWalls = function (neighbor) {
  if ((neighbor.x_coordinate - this.x_coordinate) == 1){
    this.Walls["right"].show = false;
    neighbor.Walls["left"].show = false;
  }
  else if ((neighbor.x_coordinate - this.x_coordinate) == -1) {
    this.Walls["left"].show = false;
    neighbor.Walls["right"].show = false;
  }
  else if ((this.y_coordinate - neighbor.y_coordinate) == 1) {
    this.Walls["top"].show = false;
    neighbor.Walls["bottom"].show = false;
  }
  else if ((this.y_coordinate - neighbor.y_coordinate) == -1){
    this.Walls["bottom"].show = false;
    neighbor.Walls["top"].show = false;
  }
};

File: Python/MazeObjects/Cell.py
import pygame
from MazeObjects.Colors import Color
from MazeObjects.Wall import Wall
import random


class Cell:
    def __init__(self, x_coordinate, y_coordinate, width):
        """
        Creates an instance of the cell class
        :param x_coordinate:
        :param y_coordinate:
        :param width: How wide and tall the cell would be.
        """
        self.x_coordinate = x_coordinate
        self.y_coordinate = y_coordinate
        self.__width = width
        self.visited = False
        self.walls = self.build_walls()
        self.__palette = Color()

    def build_walls(self):
        """
        Defines each wall of the cell
        :return: a collection of walls for the cell
        """
        x = self.x_coordinate * self.__width
        y = self.y_coordinate * self.__width

        top = Wall((x, y), ((x + self.__width), y))
        bottom = Wall((x, y + self.__width), ((x + self.__width), y + self.__width))
        left = Wall((x, y), (x, (y + self.__width)))
        right = Wall(((x + self.__width), y), ((x + self.__width), (y + self.__width)))

        walls = {'top': top, 'bottom': bottom, 'left': left, 'right': right}

        return walls

    def draw(self, surface):
        """
        draws a square shaped cell on the canvas/screen/whatever
        Each cell is composed of four sides. Each side is an individual line.
        :return: nothing
        """

        for wall in self.walls:
            if self.walls[wall].show:
                self.walls[wall].draw(surface)

        if self.visited:
            self.highlight(surface, self.__palette.purple, 50)

    def highlight(self, surface, color, alpha):
        """
        Fills the space occupied by the cell with the specified color
        :param surface: The canvas where the cell is drawn
        :param color: The color to be used to fill the cell
        :param alpha: The alpha level to be used along with the specified color
        :return: None
        """
        x = self.x_coordinate * self.__width
        y = self.y_coordinate * self.__width
        rectangle = pygame.Surface((self.__width, self.__width))
        rectangle.fill(color)
        rectangle.set_alpha(alpha)
        surface.blit(rectangle, (x, y))

    @staticmethod
    def get_cell(x_coordinate: int, y_coordinate: int, column_count: int=10)->int:
        """
        Gets the index of a cell from the given set of coordinates
        :param (int) x_coordinate: the x coordinate of the base cell
        :param (int) y_coordinate: the y coordinate of the base cell
        :param (int) column_count: the number of columns that the grid has
        :return: (int) the index of the specific cell in the grid
        """
        if (x_coordinate < 0) or (y_coordinate < 0) or (x_coordinate > column_count - 1) \
                or (y_coordinate > column_count - 1):
            return -1

        # print("index: {} ".format(x_coordinate + y_coordinate * column_count))
        return x_coordinate + (y_coordinate * column_count)

    def get_neighbor(self, cells, width):
        indexes = [self.get_cell(self.x_coordinate, self.y_coordinate - 1, width),  # top side
                   self.get_cell(self.x_coordinate, self.y_coordinate + 1, width),  # bottom side
                   self.get_cell(self.x_coordinate - 1, self.y_coordinate, width),  # left side
                   self.get_cell(self.x_coordinate + 1, self.y_coordinate, width)]  # right side

        neighbors = []

        for index in indexes:
            if index != -1 and cells[index].visited is False:
                neighbors.append(cells[index])

        if neighbors:
            v = random.choice(neighbors)
            return cells.index(v)
